modules/gapi/src/executor/gtbbexecutor.cpp

```diff
@@ -0,0 +1,470 @@
+                        //   (There was no active work before (i.e. root->ref_count() was == 1))
+                        auto wake_master = (master_was_active == is_tbb_work_present::no) ?
+                                wake_tbb_master::yes : wake_tbb_master::no;
+                        master_sleep_lock.get().unlock(wake_master);
+                    }
+                };
+
+                auto& body = util::get<tile_node::async_task_body>(node->task_body).body;
+                body(std::move(callback), node->total_order_index);
+            }
+
+            ctx.executed++;
+            // reset dependecy_count to initial state to simplify re-execution of the same graph
+            node->dependency_count = node->dependencies;
+
+            return result;
+        }
+    };
+}
+} // namespace detail
+}}}  // namespace cv::gimpl::parallel
+
+void cv::gimpl::parallel::execute(prio_items_queue_t& q)
+{
+    // get the reference to current task_arena (i.e. one we are running in)
+#if TBB_INTERFACE_VERSION > 9002
+    using attach_t = tbb::task_arena::attach;
+#else
+    using attach_t = tbb::internal::attach;
+#endif
+
+    tbb::task_arena arena{attach_t{}};
+    execute(q, arena);
+}
+
+void cv::gimpl::parallel::execute(prio_items_queue_t& q, tbb::task_arena& arena)
+{
+    using namespace detail;
+    graph::exec_ctx ctx{arena, q};
+
+    arena.execute(
+        [&](){
+            // Passed in queue is assumed to contain starting tasks, i.e. ones with no (or resolved) dependencies
+            auto num_start_tasks = q.size();
+
+            // TODO: use recursive spawning and task soft affinity for faster task distribution
+            // As graph is starting and no task has been spawned yet
+            // assert_graph_is_running(root) will not hold, so spawn without assert
+            tasking::batch_spawn(num_start_tasks, ctx.root.get(), graph::task_body{ctx}, /* assert_graph_is_running*/false);
+
+            std::chrono::high_resolution_clock timer;
+
+            auto tbb_work_done   = [&ctx](){ return 1 == ctx.root->ref_count(); };
+            auto async_work_done = [&ctx](){ return 0 == ctx.async_tasks.count; };
+            do {
+               // First participate in execution of TBB graph till there are no more ready tasks.
+               ctx.root->wait_for_all();
+
+               if (!async_work_done()) { // Bypass waiting on cv if there no async work to do
+                   // FIXME: use TBB resumable tasks here to avoid blocking TBB thread here
+                   auto start = timer.now();
+                   std::unique_lock<std::mutex> lk(ctx.async_tasks.mtx);
+                   // then wait (probably by sleeping) until all async tasks are completed or new TBB tasks are created.
+                   ctx.async_tasks.cv.wait(lk, [&]{return async_work_done() || !tbb_work_done() ;});
+
+                   using namespace std::chrono;
```

Comment:
M b it is reasonable to use this before `timer` is declared

Reply:
moved